ns/nsenterclass.man
.lf 1 nsenterclass.man
.\" @(#)nsenterclass.man,v 1.6 2002/08/23 12:48:51 CERN IT-PDP/DM Jean-Philippe Baud
.\" Copyright (C) 2000-2002 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH NSENTERCLASS 1 "2002/08/23 12:48:51" CASTOR "Cns Administrator Commands"
.SH NAME
nsenterclass \- define a new file class
.SH SYNOPSIS
.B nsenterclass
.BI --id " classid"
.BI --name " class_name"
[
.BI -h " name_server"
] [
.BI --flags " flags"
] [
.BI --gid " class_gid"
] [
.BI --group " class_group"
] [
.BI --maxdrives " n"
] [
.BI --maxfilesize " n"
] [
.BI --maxsegsize " n"
] [
.BI --migr_interval " n"
] [
.BI --minfilesize " n"
] [
.BI --mintime " n"
] [
.BI --nbcopies " n"
] [
.BI --retenp_on_disk " n"
] [
.BI --tppools " pool1:pool2..."
] [
.BI --uid " class_uid"
] [
.BI --user " class_user"
]
.SH DESCRIPTION
.B nsenterclass
defines a new file class.
.LP
This command requires ADMIN privilege in the Cupv database.
.SH OPTIONS
.TP
.BI -h " name_server"
.TP
.BI --flags " flags"
.TP
.BI --gid " class_gid"
The class is restricted to this group gid.
.TP
.BI --id " classid"
The class number must be strictly positive.
.TP
.BI --maxdrives " n"
Specifies the number of drives which can be used in parallel for a migration.
.TP
.BI --maxfilesize " n"
Specifies the maximum file size. There is no limit by default.
.TP
.BI --maxsegsize " n"
Specifies the maximum segment size. There is no limit by default.
.TP
.BI --migr_interval " n"
A new migration will be started if at least
.I n
seconds have elapsed since the last migration.
.TP
.BI --minfilesize " n"
Specifies the minimum file size. The default value is 0.
.TP
.BI --mintime " n"
A file will not be migrated unless at least
.I n
seconds have elapsed since the last update.
.TP
.BI --name " class_name"
The class name must be at most CA_MAXCLASNAMELEN characters long.
.TP
.BI --nbcopies " n"
Specifies the number of copies for a file. Each copy is written to a different
tape pool.
.TP
.BI --retenp_on_disk " n"
Specifies the maximum retention period (in seconds) for a file on disk.
The retention period can also be set to AS_LONG_AS_POSSIBLE or INFINITE_LIFETIME.
If zero, the file is purged immediately after migration.
Default is AS_LONG_AS_POSSIBLE, i.e. purged when disk space is needed.
.TP
.BI --tppools " pool1:pool2..."
Specifies the tape pools to be used for migration. The number of tape pools
must be at least as big as the number of copies.
Pool names must be at most CA_MAXPOOLNAMELEN characters long.
.TP
.BI --uid " class_uid"
The class is restricted to this user uid.
.SH EXAMPLES
.nf
.ft CW
nsenterclass --id 2 --name user --maxd 2 --migr 60 --nbco 1 --tppools default
.ft
.fi
.SH EXIT STATUS
This program returns 0 if the operation was successful or >0 if the operation
failed.
.SH FILES
.TP
.B /etc/group
.TP
.B /etc/passwd
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR Cns_enterclass(3) ,
.B Cupvlist(1)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


ns/nsenterclass.c
/*
 * Copyright (C) 2000-2001 by CERN/IT/PDP/DM
 * All rights reserved
 */

#ifndef lint
static char sccsid[] = "@(#)nsenterclass.c,v 1.7 2001/03/08 07:04:42 CERN IT-PDP/DM Jean-Philippe Baud";
#endif /* not lint */

/*	nsenterclass - define a new file class */
#include <errno.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include "Cgetopt.h"
#include "Cns_api.h"
#include "serrno.h"
main(argc, argv)
int argc;
char **argv;
{
	int c;
	struct Cns_fileclass Cns_fileclass;
	char *dp;
	int errflg = 0;
	struct group *gr;
	static struct Coptions longopts[] = {
		{"gid", REQUIRED_ARGUMENT, 0, OPT_CLASS_GID},
		{"group", REQUIRED_ARGUMENT, 0, OPT_CLASS_GROUP},
		{"id", REQUIRED_ARGUMENT, 0, OPT_CLASS_ID},
		{"name", REQUIRED_ARGUMENT, 0, OPT_CLASS_NAME},
		{"uid", REQUIRED_ARGUMENT, 0, OPT_CLASS_UID},
		{"user", REQUIRED_ARGUMENT, 0, OPT_CLASS_USER},
		{"flags", REQUIRED_ARGUMENT, 0, OPT_FLAGS},
		{"maxdrives", REQUIRED_ARGUMENT, 0, OPT_MAX_DRV},
		{"maxfilesize", REQUIRED_ARGUMENT, 0, OPT_MAX_FSZ},
		{"maxsegsize", REQUIRED_ARGUMENT, 0, OPT_MAX_SSZ},
		{"migr_interval", REQUIRED_ARGUMENT, 0, OPT_MIGR_INTV},
		{"minfilesize", REQUIRED_ARGUMENT, 0, OPT_MIN_FSZ},
		{"mintime", REQUIRED_ARGUMENT, 0, OPT_MIN_TIME},
		{"nbcopies", REQUIRED_ARGUMENT, 0, OPT_NBCOPIES},
		{"retenp_on_disk", REQUIRED_ARGUMENT, 0, OPT_RETENP_DISK},
		{"tppools", REQUIRED_ARGUMENT, 0, OPT_TPPOOLS},
		{0, 0, 0, 0}
	};
	int nbtppools;
	char *p;
	struct passwd *pwd;
	char *q;
	char *server = NULL;

	memset (&Cns_fileclass, 0, sizeof(struct Cns_fileclass));
	Cns_fileclass.retenp_on_disk = AS_LONG_AS_POSSIBLE;
	Copterr = 1;
	Coptind = 1;
	while ((c = Cgetopt_long (argc, argv, "h:", longopts, NULL)) != EOF) {
		switch (c) {
		case OPT_CLASS_GID:
			if ((Cns_fileclass.gid = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid class_gid %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_CLASS_GROUP:
			if ((gr = getgrnam (Coptarg)) == NULL) {
				fprintf (stderr,
				    "invalid class_group: %s\n", Coptarg);
				errflg++;
			} else
				Cns_fileclass.gid = gr->gr_gid;
			break;
		case OPT_CLASS_ID:
			if ((Cns_fileclass.classid = strtol (Coptarg, &dp, 10)) <= 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid classid %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_CLASS_NAME:
			if (strlen (Coptarg) > CA_MAXCLASNAMELEN) {
				fprintf (stderr,
				    "class name too long: %s\n", Coptarg);
				errflg++;
			} else
				strcpy (Cns_fileclass.name, Coptarg);
			break;
		case OPT_CLASS_UID:
			if ((Cns_fileclass.uid = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid class_uid %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_CLASS_USER:
			if ((pwd = getpwnam (Coptarg)) == NULL) {
				fprintf (stderr,
				    "invalid class_user: %s\n", Coptarg);
				errflg++;
			} else
				Cns_fileclass.uid = pwd->pw_uid;
			break;
		case OPT_FLAGS:
			if ((Cns_fileclass.flags = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid flags %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MAX_DRV:
			if ((Cns_fileclass.maxdrives = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid maxdrives %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MAX_FSZ:
			if ((Cns_fileclass.max_filesize = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid maxfilesize %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MAX_SSZ:
			if ((Cns_fileclass.max_segsize = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid maxsegsize %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MIGR_INTV:
			if ((Cns_fileclass.migr_time_interval = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid migr_interval %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MIN_FSZ:
			if ((Cns_fileclass.min_filesize = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid minfilesize %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_MIN_TIME:
			if ((Cns_fileclass.mintime_beforemigr = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid mintime %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_NBCOPIES:
			if ((Cns_fileclass.nbcopies = strtol (Coptarg, &dp, 10)) < 0 ||
			    *dp != '\0') {
				fprintf (stderr,
				    "invalid nbcopies %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_RETENP_DISK:
			if (strcmp (Coptarg, "AS_LONG_AS_POSSIBLE") == 0)
				Cns_fileclass.retenp_on_disk = AS_LONG_AS_POSSIBLE;
			else if (strcmp (Coptarg, "INFINITE_LIFETIME") == 0)
				Cns_fileclass.retenp_on_disk = INFINITE_LIFETIME;
			else if ((Cns_fileclass.retenp_on_disk =
			    strtol (Coptarg, &dp, 10)) < 0 || *dp != '\0') {
				fprintf (stderr,
				    "invalid retenp_on_disk %s\n", Coptarg);
				errflg++;
			}
			break;
		case OPT_TPPOOLS:
			nbtppools = 0;
			p = strtok (Coptarg, ":");
			while (p) {
				if (strlen (p) > CA_MAXPOOLNAMELEN) {
					fprintf (stderr,
					    "invalid tppools %s\n", Coptarg);
					errflg++;
				} else
					nbtppools++;
				if (p = strtok (NULL, ":")) *(p - 1) = ':';
			}
			if (errflg) break;
			Cns_fileclass.nbtppools = nbtppools;
			if ((q = calloc (nbtppools, CA_MAXPOOLNAMELEN+1)) == NULL) {
				fprintf (stderr, "nsenterclass %s: %s\n",
				    Cns_fileclass.name, sstrerror(ENOMEM));
				errflg++;
				break;
			}
			Cns_fileclass.tppools = q;
			p = strtok (Coptarg, ":");
			while (p) {
				strcpy (q, p);
				q += (CA_MAXPOOLNAMELEN+1);
				p = strtok (NULL, ":");
			}
			break;
		case 'h':
			server = Coptarg;
			break;
		case '?':
			errflg++;
			break;
		default:
			break;
		}
	}
	if (Coptind < argc || Cns_fileclass.name == NULL) {
		errflg++;
	}
	if (errflg) {
		fprintf (stderr, "usage: %s %s %s %s %s %s", argv[0],
		    "--id classid --name class_name [-h name_server] [--flags flags]\n",
		    "[--gid class_gid] [--group class_group] [--maxdrives n] [--maxfilesize n]\n",
		    "[--maxsegsize n] [--migr_interval n] [--minfilesize n] [--mintime n]\n",
		    "[--nbcopies n] [--retenp_on_disk n] [--tppools pool1:pool2...]\n",
		    "[--uid class_uid] [--user class_user]\n");
		exit (USERR);
	}

	if (Cns_enterclass (server, &Cns_fileclass) < 0) {
		fprintf (stderr, "nsenterclass %s: %s\n", Cns_fileclass.name,
		    sstrerror(serrno));
		exit (USERR);
	}
	exit (0);
}
